letsencrypt/cli.py

```diff
@@ -283,22 +282,61 @@ def run(args, config, plugins):  # pylint: disable=too-many-branches,too-many-lo
         lineage.update_all_links_to(lineage.latest_common_version())
         # TODO: Check return value of save_successor
         # TODO: Also update lineage renewal config with any relevant
-        #       configuration values from this attempt?
-        le_client.deploy_certificate(
-            domains, lineage.privkey, lineage.cert, lineage.chain)
-        display_ops.success_renewal(domains)
+        #       configuration values from this attempt? - YES
     else:
         # TREAT AS NEW REQUEST
         lineage = le_client.obtain_and_enroll_certificate(
-            domains, authenticator, installer, plugins)
+            domains, le_client.dv_auth, le_client.installer, plugins)
         if not lineage:
-            return "Certificate could not be obtained"
-        # TODO: This treats the key as changed even when it wasn't
-        # TODO: We also need to pass the fullchain (for Nginx)
-        le_client.deploy_certificate(
-            domains, lineage.privkey, lineage.cert, lineage.chain)
-        le_client.enhance_config(domains, args.redirect)
+            raise errors.Error("Certificate could not be obtained")
+
+    return lineage
+
+
+# TODO: Make run as close to auth + install as possible
+# Possible difficulties: args.csr was hacked into auth
+def run(args, config, plugins):  # pylint: disable=too-many-branches,too-many-locals
+    """Obtain a certificate and install."""
+    # Begin authenticator and installer setup
+    if args.configurator is not None and (args.installer is not None or
+                                          args.authenticator is not None):
+        return ("Either --configurator or --authenticator/--installer"
+                "pair, but not both, is allowed")
+
+    if args.authenticator is not None or args.installer is not None:
+        installer = display_ops.pick_installer(
+            config, args.installer, plugins)
+        authenticator = display_ops.pick_authenticator(
+            config, args.authenticator, plugins)
+    else:
+        # TODO: this assumes that user doesn't want to pick authenticator
+        #       and installer separately...
+        authenticator = installer = display_ops.pick_configurator(
+            config, args.configurator, plugins)
+
+    if installer is None or authenticator is None:
+        return "Configurator could not be determined"
+    # End authenticator and installer setup
+
+    domains = _find_domains(args, installer)
+
+    # TODO: Handle errors from _init_le_client?
+    le_client = _init_le_client(args, config, authenticator, installer)
+
+    try:
+        lineage = _auth_from_domains(le_client, config, domains, plugins)
+    except errors.Error as err:
+        return str(err)
```

Comment:
I am not sure if this is right... @bmw?  I am guessing I should have just let this float up and it would be handled by your code?